File: scripts/env-ports.js
#!/usr/bin/env node
/**
 * Load port numbers from a .env file. Use from scripts or programmatically.
 *
 * Node usage:
 *   const { getPortsFromEnv } = require('./scripts/env-ports.js');
 *   const ports = getPortsFromEnv('/path/to/project/.env', { EXPRESS_PORT: 3000, VITE_DEV_PORT: 5173 });
 */

const path = require('path');
const fs = require('fs');

/**
 * Load .env from envPath (or cwd), parse it, and return port values for the given keys.
 * @param {string} [envPath] - Path to .env file. If omitted, uses .env in process.cwd().
 * @param {Record<string, number>} defaults - Map of env var name -> default port if missing/invalid.
 * @returns {Record<string, number>} Map of env var name -> port number (integer).
 */
function getPortsFromEnv(envPath, defaults = {}) {
  const resolved = envPath
    ? path.resolve(envPath)
    : path.join(process.cwd(), '.env');

  if (fs.existsSync(resolved)) {
    const content = fs.readFileSync(resolved, 'utf8');
    const env = {};
    for (const line of content.split('\n')) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith('#')) continue;
      const eq = trimmed.indexOf('=');
      if (eq === -1) continue;
      const key = trimmed.slice(0, eq).trim();
      let val = trimmed.slice(eq + 1).trim();
      if (val.startsWith('"') && val.endsWith('"')) val = val.slice(1, -1);
      if (val.startsWith("'") && val.endsWith("'")) val = val.slice(1, -1);
      env[key] = val;
    }
    for (const key of Object.keys(defaults)) {
      const raw = env[key];
      const num = raw != null ? parseInt(raw, 10) : NaN;
      env[key] = Number.isInteger(num) && num > 0 && num < 65536 ? num : defaults[key];
    }
    return Object.fromEntries(
      Object.keys(defaults).map((k) => [k, env[k] ?? defaults[k]])
    );
  }

  return { ...defaults };
}

/**
 * Convenience: load .env with dotenv then return ports (mutates process.env).
 * Use when you want process.env to be set (e.g. for child processes).
 */
function loadEnvAndGetPorts(envPath, defaults = {}) {
  const resolved = envPath
    ? path.resolve(envPath)
    : path.join(process.cwd(), '.env');
  try {
    require('dotenv').config({ path: resolved });
  } catch (_) {}
  const result = {};
  for (const [key, def] of Object.entries(defaults)) {
    const raw = process.env[key];
    const num = raw != null ? parseInt(raw, 10) : NaN;
    result[key] = Number.isInteger(num) && num > 0 && num < 65536 ? num : def;
  }
  return result;
}

module.exports = { getPortsFromEnv, loadEnvAndGetPorts };

File: scripts/env_ports.py
#!/usr/bin/env python3
"""
Load port numbers from a .env file. Use from scripts or other Python code.

Python usage:
    from env_ports import get_ports_from_env
    ports = get_ports_from_env("/path/to/project/.env", defaults={"EXPRESS_PORT": 3000, "VITE_DEV_PORT": 5173})
    port = ports["EXPRESS_PORT"]
"""

import os
from pathlib import Path


def _parse_env_file(path: Path) -> dict[str, str]:
    """Parse .env file into key -> value (string). No variable expansion."""
    env = {}
    if not path.exists():
        return env
    with open(path, encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            eq = line.find("=")
            if eq == -1:
                continue
            key = line[:eq].strip()
            val = line[eq + 1 :].strip()
            if (val.startswith('"') and val.endswith('"')) or (
                val.startswith("'") and val.endswith("'")
            ):
                val = val[1:-1]
            env[key] = val
    return env


def get_ports_from_env(
    env_path: str | Path | None = None,
    defaults: dict[str, int] | None = None,
) -> dict[str, int]:
    """
    Load .env from env_path (or cwd), parse it, and return port values for the given keys.

    :param env_path: Path to .env file. If None, uses .env in current working directory.
    :param defaults: Map of env var name -> default port if missing/invalid.
    :return: Map of env var name -> port number (int).
    """
    if defaults is None:
        defaults = {}
    resolved = (
        Path(env_path).resolve()
        if env_path is not None
        else Path.cwd() / ".env"
    )
    env = _parse_env_file(resolved)
    result = {}
    for key, default in defaults.items():
        raw = env.get(key) or os.environ.get(key)
        try:
            num = int(raw) if raw is not None else default
        except (TypeError, ValueError):
            num = default
        if not (0 < num < 65536):
            num = default
        result[key] = num
    return result


def load_dotenv_and_get_ports(
    env_path: str | Path | None = None,
    defaults: dict[str, int] | None = None,
) -> dict[str, int]:
    """
    Load .env with python-dotenv (if available), then return ports.
    Mutates os.environ so child processes see the vars.
    """
    if defaults is None:
        defaults = {}
    resolved = (
        Path(env_path).resolve()
        if env_path is not None
        else Path.cwd() / ".env"
    )
    try:
        from dotenv import load_dotenv
        load_dotenv(resolved)
    except ImportError:
        pass
    return get_ports_from_env(resolved, defaults)
